lib/ocpp/v201/profile.cpp

```diff
@@ -0,0 +1,684 @@
+                if (seconds_to_go_back < 0) {
+                    seconds_to_go_back += SECONDS_PER_DAY * DAYS_PER_WEEK;
+                }
+                seconds_to_go_forward = SECONDS_PER_DAY * DAYS_PER_WEEK;
+                break;
+            default:
+                EVLOG_error << "Invalid RecurrencyKindType: " << static_cast<int>(in_profile.recurrencyKind.value());
+                break;
+            }
+
+            start = std::move(DateTime(now_tp - seconds(seconds_to_go_back)));
+
+            while (start <= end) {
+                start_times.push_back(start);
+                start = DateTime(start.to_time_point() + seconds(seconds_to_go_forward));
+            }
+        }
+        break;
+    case ChargingProfileKindEnum::Relative:
+        // if there isn't a session start then assume the session starts now
+        if (in_session_start) {
+            start = floor_seconds(in_session_start.value());
+        }
+        start_times.push_back(start);
+        break;
+    }
+    return start_times;
+}
+
+/// \brief calculate the start times for the schedule period
+/// \param in_now the current date and time
+/// \param in_end the end of the composite schedule
+/// \param in_session_start optional when the charging session started
+/// \param in_profile the charging profile
+/// \param in_period_index the schedule period index
+/// \return the list of start times
+std::vector<period_entry_t> calculate_profile_entry(const DateTime& in_now, const DateTime& in_end,
+                                                    const std::optional<DateTime>& in_session_start,
+                                                    const ChargingProfile& in_profile, std::uint8_t in_period_index) {
+    std::vector<period_entry_t> entries;
+
+    if (in_period_index >= in_profile.chargingSchedule.front().chargingSchedulePeriod.size()) {
+        EVLOG_error << "Invalid schedule period index [" << static_cast<int>(in_period_index)
+                    << "] (too large) for profile " << in_profile.id;
+    } else {
+        const auto& this_period = in_profile.chargingSchedule.front().chargingSchedulePeriod[in_period_index];
+
+        if ((in_period_index == 0) && (this_period.startPeriod != 0)) {
+            // invalid profile - first period must be 0
+            EVLOG_error << "Invalid schedule period index [0] startPeriod " << this_period.startPeriod
+                        << " for profile " << in_profile.id;
+        } else if ((in_period_index > 0) &&
+                   (in_profile.chargingSchedule.front().chargingSchedulePeriod[in_period_index - 1].startPeriod >=
+                    this_period.startPeriod)) {
+            // invalid profile - periods must be in order and with increasing startPeriod values
+            EVLOG_error << "Invalid schedule period index [" << static_cast<int>(in_period_index) << "] startPeriod "
+                        << this_period.startPeriod << " for profile " << in_profile.id;
```

Comment:
I like the logging! Is this type of invalid profile not rejected by the `validate_profile` method? This should definitely be caught before we accept a submitted charge profile.

Reply:
That's a good question. I'd have to investigate.